### ISSUE TITLE
InvalidReturnStatement erroneously emitted for a Generator

### ISSUE THREAD

**Issue by TysonAndre** [vimeo/psalm#485]
Expected: Psalm should not warn. Because `yield` exists, the return value is a generator, and Generator->getReturn() is null.

```php
<?php

/**
 * @return Generator
 */
function generator2() : Generator {
    if (rand(0,1)) {
        return;  // Erroneous "InvalidReturnStatement - src/generator.php:8:9 - Empty return statement is not expected in generator2 "
    }
    yield 2;
}
```

**Comment by TysonAndre** [comment]
A more comprehensive test case:

```php
<?php

/**
 * @return Generator
 */
function generator2() : Generator {
    if (rand(0,1)) {
        return;
    }
    yield 2;
}

/**
 * @return Generator
 * @suppress InvalidNullableReturnType (should continue to emit InvalidReturnStatement)
 */
function notagenerator() : Generator {
    if (rand(0,1)) {
        return;
    }
    return generator2();
}
```

**Comment by muglug** [comment]
I've fixed with the same basic hack you suggested – any reason why it wouldn't work?

**Comment by muglug** [comment]
Ah ok, I guess the solution here is to check for `yield` statements at the scanning stage.
